internal/venice-common/src/main/java/com/linkedin/venice/pubsub/manager/TopicMetadataFetcher.java

```diff
@@ -0,0 +1,673 @@
+      TopicManagerStats.recordLatency(
+          topicManagerStats,
+          TopicManagerStats.OCCURRENCE_LATENCY_SENSOR_TYPE.GET_PARTITION_LATEST_OFFSETS,
+          System.currentTimeMillis() - startTime);
+      Long offset = offsetMap.get(pubSubTopicPartition);
+      if (offset == null) {
+        // This should never happen; if it does, it's a bug in the PubSubConsumerAdapter implementation
+        throw new VeniceException("Got null as latest offset for: " + pubSubTopicPartition);
+      }
+      return offset;
+    } finally {
+      releaseConsumer(pubSubConsumerAdapter);
+    }
+  }
+
+  long getLatestOffsetWithRetries(PubSubTopicPartition pubSubTopicPartition, int retries) {
+    VeniceException lastException = DUMMY_PUBSUB_OP_TIMEOUT_EXCEPTION;
+    retries = Math.max(1, retries);
+    int attempt = 0;
+    while (attempt < retries) {
+      try {
+        return getLatestOffset(pubSubTopicPartition);
+      } catch (PubSubTopicDoesNotExistException | PubSubOpTimeoutException e) {
+        lastException = e;
+        attempt++;
+        LOGGER.warn(
+            "Failed to get the latest offset in attempt: {}/{} for: {}.",
+            attempt,
+            retries,
+            pubSubTopicPartition,
+            e);
+      }
+    }
+    throw lastException;
+  }
+
+  CompletableFuture<Long> getLatestOffsetWithRetriesAsync(PubSubTopicPartition pubSubTopicPartition, int retries) {
+    return CompletableFuture
+        .supplyAsync(() -> getLatestOffsetWithRetries(pubSubTopicPartition, retries), threadPoolExecutor);
+  }
+
+  long getLatestOffsetCached(PubSubTopicPartition pubSubTopicPartition) {
+    long now = System.nanoTime();
+    ValueAndExpiryTime<Long> cachedValue;
+    try {
+      cachedValue = latestOffsetCache.computeIfAbsent(pubSubTopicPartition, k -> {
+        long latestOffset =
+            getLatestOffsetWithRetries(pubSubTopicPartition, DEFAULT_MAX_RETRIES_FOR_POPULATING_TMD_CACHE_ENTRY);
+        return new ValueAndExpiryTime<>(latestOffset, now + cachedEntryTtlInNs);
+      });
+    } catch (PubSubTopicDoesNotExistException | PubSubOpTimeoutException e) {
+      LOGGER.error("Failed to get end offset for topic-partition: {}", pubSubTopicPartition, e);
+      return StatsErrorCode.LAG_MEASUREMENT_FAILURE.code;
+    }
+    populateCacheWithLatestOffset(pubSubTopicPartition);
+    return cachedValue.getValue();
+  }
+
+  // load the cache with the latest offset
+  void populateCacheWithLatestOffset(PubSubTopicPartition pubSubTopicPartition) {
+    long now = System.nanoTime();
+    // if there is no entry in the cache or the entry has expired, we will try to populate the cache
+    ValueAndExpiryTime<Long> cachedValue = latestOffsetCache.putIfAbsent(
+        pubSubTopicPartition,
+        new ValueAndExpiryTime<>((long) StatsErrorCode.LAG_MEASUREMENT_FAILURE.code, now + cachedEntryTtlInNs));
+    if (cachedValue == null || (cachedValue.getExpiryTimeNs() <= now && cachedValue.tryAcquireUpdateLock())) {
```

Comment:
Nit, do you think we could switch the order here in the if condition to something below. I am thinking of a small inefficiency case here, where two threads get the `cachedValue` at the same time, after thread 1 passed `cachedValue.getExpiryTimeNs() <= now` condition, now thread 2 got the lock, finished the `updateValue(latestOffset)`. Now thread 1 continue with getting the lock and updating value etc, but it does't have to (as value has just been updated by thread 2), if it `getExpiryTimeNs` after   `tryAcquireUpdateLock`



`if (cachedValue == null || (cachedValue.tryAcquireUpdateLock() && cachedValue.getExpiryTimeNs() <= now)`

Reply:
It's probably a benign race (just efficiency rather than correctness issue), but do we want to just leave the race there due to the overhead you mentioned? On the other hand, there is not much overhead, if we look at the way of acquiring the "lock", it's not really a lock and it's non blocking + no need to release, is that right?

```
    boolean tryAcquireUpdateLock() {
      return isUpdateInProgress.compareAndSet(false, true);
    }
```

Reply:
I believe there are tradeoffs to consider with each of these options:

1) `(cachedValue.tryAcquireUpdateLock() && cachedValue.getExpiryTimeNs() <= now)`
   With this option, for each request, we attempt to acquire the lock (update the atomic variable). If multiple threads attempt to update the atomic variable simultaneously, we incur a performance penalty. Additionally, with this approach, we must ensure that the lock is released (setting `isUpdateInProgress` to `false`) if the value is not expired, so that subsequent callers can obtain the lock.

2) `(cachedValue.getExpiryTimeNs() <= now && cachedValue.tryAcquireUpdateLock())`
   In this case, we acquire the lock only when necessary and hence reduced the chances of multiple threads trying to update atomic var at the same time. The downside is that occasionally we may encounter the race condition you mentioned.

I believe option (2) is simpler and relatively better than option (1).

We can further improve option (2) by skipping the update (although this means we will be checking the expiry time twice: once before grabbing the lock and once after obtaining it) if the value is not expired.

Reply:
Do you think we can make this improvement in this pr?

> We can further improve option (2) by skipping the update (although this means we will be checking the expiry time twice: once before grabbing the lock and once after obtaining it) if the value is not expired.

Reply:
Yeah sure, I'll update the PR. Thanks!

Reply:
Updated. Could you please take another look? Thanks!